### DISCUSSION TITLE
Preview mode not working in production

### DISCUSSION THREAD

**Discussion by huttotw** [vercel/next.js#12898] [Help] [answered]
I have tested preview mode locally and everything works fine. However, in production for some reason the cookies are not being set.

In production, the site is served over HTTPS, but the connection from the load balancer to the Next server is not over HTTPS, is that a problem?

Another note, when I curl my preview API endpoint `/api/preview/activate`, I get the Set-Cookie headers back.

```
set-cookie: __prerender_bypass=5b870eaf3a8e3870d9bfb89a95bda077; Max-Age=900; Path=/; HttpOnly; Secure; SameSite=None
set-cookie: __next_preview_data=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkYXRhIjoiMGFkMDg3MmZiNzM2ZTBlMGIzNzAzMDA1MTUzMTZjNjUzZTVhOWQ2ZjcwYjMwZDRlMTI0NmFlZWU1YzMwNDc1MDFmMGM4MWZmNzA3OWZkOGIwZjUwMzJlMjgxZTEyODI4NzRhMjRmYmU2MDRmZTgwYWU3NzMwNzA0MDZiNDA0NDQ3ZWRkMzIyNjQ5MDYzZTUyYWEzZTRlMzA3NWNiOGY5ZWY1MTUzMjMwMDljOTBhNzQ2YTM4MzgzM2RhMjVjYTI3MDI0YSIsImlhdCI6MTU4OTQ5MDk5MiwiZXhwIjoxNTg5NDkxODkyfQ.Pp3KQaP1bffTp7pJUSElrHCcCeWbt8zdOORZwb0IweY; Max-Age=900; Path=/; HttpOnly; Secure; SameSite=None
```

```js
const url = require('url');

export default (req, res) => {
  const params = url.parse(req.url, true).query;
  res.setPreviewData(
    {},
    {
      maxAge: 60 * 15, // The preview mode cookies expire in 15 minutes
    }
  );
  res.writeHead(307, { Location: params.redirectUri });
  res.end();
  res.finished = true;
};

```

It is entirely possible I have a fundamental misunderstanding of cookies. Any advice would be appreciated, I've been staring at this for hours.

**Comment by huttotw** [comment]
Okay, so taking the redirect has allowed the cookies to be set, but on the next page load they are removed. Something very strange is going on here.

**Comment by huttotw** [comment, marked as answer]
Found the issue, build IDs are important!

I had the build ID hardcoded since we build the site on each server during deployment. Making these dynamic, for instance calculating a checksum of your project should do the trick.

**Reply by mia01** [reply]
@huttotw I'm having this same issue - we are running next js on multiple kubernetes pods - is there any link or documentation which explains the next js build ids? do they have to be the same on all pods?

**Comment by gitabhisingh** [comment]
I am building static pages using Contentful CMS and I have issues using preview mode on prod build(localhost)
After curl I found out that cookies are being set but in **SameSite=None** mode. Is that the issue?
I am using Next JS v9.5.2

```
 HTTP/1.1 200 OK
HTTP/1.1 200 OK
< Set-Cookie: __prerender_bypass=e00e4ba3b660386df18eea2687f9b75; Path=/; HttpOnly; Secure; SameSite=None
Set-Cookie: __prerender_bypass=e00e4ba3b660386ef18eea2687f9b75; Path=/; HttpOnly; Secure; SameSite=None
< Set-Cookie: __next_preview_data=eyJhbGciOiJIUzI1NiIsInR5cI6IkpXVCJ9.eyJkYXRhIjoiYTU0NWQyN2RkZTM1ZTBiNjhiNzE3MTBkOWMxZjBmMGVlZDU3ZjEwMGQ5ZmExZTIxYmRjMDliMDY4Yjc5ODA1MWNlYzdhYWU1OWY5MzY1MTI5Y2RiMTNlZTA2OTM3M2ZhYjVkYjQ3NjYxYjdlMTkzMThhNmM4N2MxY2MzZTI3NDVhND4YjI0ZjllODgxN2I2OWYxODdhMTFhYjE0YzUyNzE0ZDQ0NDI4OTI1MWVhOTg1MDdkNWZlMDZkNThiNmUzMzI2MSIsImlhdCI6MTYwMzQ2NzA2Nn0.V4IpUW7V_goPgiJ5mV6-axAKQcMFxqNXwUQqVp6nfA; Path=/; HttpOnly; Secure; SameSite=None
Set-Cookie: __next_preview_data=eyJhbGciOiJIUzI1NisInR5cCI6IkpXVCJ9.eyJkYXRhIjoiYTU0NWQyN2RkZTM1ZTBiNjhiNzE3MTBkOWMxZjBmMGVlZDU3ZjEwMGQ5ZmExZTIxYmRjMDliMDY4Yjc5ODA1MWNlYzdhYWU1OWY5MzY1MTI5Y2RiMTNlZTA2OTM3M2ZhYjVkYjQ3NjYxYjdlMTkzMThhNmM4N2MxY2MzZTI3NDVhNDk4YjI0ZjllODgxN2I2OWYxODdhMTFhYjE0YzUyNzE0ZDQ0NDI4OTI1MWVhOTg1MDdkNWZlMDZkNThiNmUzMzI2MSIsmlhdCI6MTYwMzQ2NzA2Nn0.V4IpUW7V_goPgiJ5mV6-axAKQcMFxqNXwWQqVp6nfA; Path=/; HttpOnly; Secure; SameSite=None
< Date: Fri, 23 Oct 2020 15:31:06 GMT
Date: Fri, 23 Oct 2020 15:31:06 GMT
< Connection: keep-alive
Connection: keep-alive
< Transfer-Encoding: chunked
Transfer-Encoding: chunked

< 
<!DOCTYPE html><html><head><meta http-equiv="Refresh" content="0; url=/posts/everything-you-want-to-know" />
    <script>window.location.href = '/posts/everything-you-want-to-know'</script>
* Connection #0 to host localhost left intact
    </head>* Closing connection 0
```

**Reply by cedricium** [reply]
Hey @shinesingh1994, were you able to solve this by any chance? I'm using Contentful as well and am also running into the same issue 😬

**Reply by gitabhisingh** [reply]
Well I found out that the cookies were not being set on http prod build, on https it should be fine. worked for me on https

**Comment by SalahAdDin** [comment]
Well, i have the same problem now when I set up the environment variable from the dashboard:
![Screen Capture_select-area_20201204004844](https://user-images.githubusercontent.com/5034215/101092530-db122f00-35ca-11eb-89de-184ce853af87.jpg)

As you can see this is a simple text variable; when I use the token in my preview, it simply sends me the `message: "Invalid token"` message.

I'm following the tutorial for doing this simple mode with **Strapi**, and it does work in local very well, but not in production.

I will test it by setting the environment variable in a proper `.env` file for production; let's see if it works properly.

**Comment by gitabhisingh** [comment]
Solved, Well I found out that the cookies were not being set on http prod build, on https it should be fine. worked for me on https

**Reply by 29hsaeed** [reply]
Hey @shinesingh1994  I have HTTPS setup on localhost using ngrok, preview mode still not working in production mode as cookies are not being set
